Build wallettemplate with JDK 11+ and OpenJFX downloaded from Maven Central

msgilligan opened the pull request:
Possible resolution to Issue #1713 

* Create new `settings-jdk11.gradle` Gradle settings file
* Update `.travis.yml` to use `-c settings-jdk11.gradle` when building with JDK11
* Update build instructions in README

msgilligan commented:
I'm almost certain the build failure is one of those intermittent integration test failures and has a good chance of running if restarted.

schildbach commented on .travis.yml:
```diff
@@ -2,13 +2,15 @@
-    # excluding wallet-template because of the unreliable JavaFX dependency
-  - gradle clean build -x :wallettemplate:compileJava
+install:
+  - sudo apt-get install -y protobuf-compiler
```
Is this a merge error, or do you intentionally want to install the protobuf debian package again?

msgilligan replied:
Shit. It's a merge error. Sorry.

msgilligan replied:
Fixed.

schildbach commented on .travis.yml:
```diff
@@ -2,13 +2,15 @@
+    - jdk: openjdk8
+      script: gradle clean build
+    - jdk: openjdk11
+      script: gradle -c settings-jdk11.gradle clean build
```
Rather than using a separate settings-jdk11.gradle, did you try using the `-x` (exclude) switch?

msgilligan replied:
I did try that, the problem is the `wallettemplate` module is using a JDK11-requiring plugin and even with the `-x` switch the build still fails.

msgilligan replied:
Actually, I have an even better solution --- stay tuned.

schildbach commented on .travis.yml:
```diff
@@ -2,13 +2,13 @@
-jdk:
-  - openjdk8
-  - openjdk11
-install: true
```
According to [Travis](https://docs.travis-ci.com/user/job-lifecycle/#skipping-the-installation-phase), the `install: true` needs to stay in if you don't have anything to do in the install phase. This is because the install phase isn't optional.

msgilligan replied:
It seemed to work OK without out, but I'll put it back.

schildbach replied:
In my experience, if you leave it out the `script` phase gets replaced with the default `gradle assemble`.

schildbach commented:
Can also rebase and squash everything?

msgilligan commented:
> Can also rebase and squash everything?

I can, but Github will do that for you automatically (if you choose the option) when you merge. Will that work for you?

schildbach commented:
> > Can also rebase and squash everything?
> 
> I can, but Github will do that for you automatically (if you choose the option) when you merge. Will that work for you?

I generally don't use the Github UI for merging. But even if I would, I couldn't due to the merge conflict.

msgilligan commented:
By the way, I'm enjoying using [sdkman](https://sdkman.io) to switch between JVMs:

```
$ sdk use java 8.0.201-oracle
$ gradle clean build
$ sdk use java 11.0.2-open
$ gradle clean build
```

msgilligan commented:
> I generally don't use the Github UI for merging. But even if I would, I couldn't due to the merge conflict.

OK, I'll fix it.

msgilligan commented:
I'm not use to doing rebase/squash from the command-line (but can learn quickly I hope) -- is this good enough?

msgilligan commented:
Actually wait for my next commit...

msgilligan commented:
@schildbach Sorry for the hassle -- I think everything is ready now.

schildbach commented on settings.gradle:
```diff
@@ -1,4 +1,6 @@
 include 'tools'
 include 'examples'
-include 'wallettemplate'
+if (JavaVersion.current().isJava11Compatible()) {
```
Unfortunately Gradle 3.4.1 (which is the version that is in Ubuntu 18.04) does not have `.isJava11Compatible()`. It would have `.isJava10Compatible()` though, in case the javafxplugin is compatible with Java 10, too.

schildbach replied:
Even Gradle 4.4.1 (the current version on Debian testing and Ubuntu 19.04) doesn't have `.isJava11Compatible()`.

msgilligan replied:
> in case the javafxplugin is compatible with Java 10, too.

Java 11 is the first version that doesn't include JavaFX by default. It's also the long-term release so it makes some sense that the plugin requires 11+.

schildbach replied:
I agree. I'm just looking for workarounds. Another workaround may be using `getMajorVersion()` and do some string foo.

schildbach replied:
Or maybe a construct like `isAtLeastGradle4.7() && isJava11Compatible()` would guard us from the exception? This would of course make the wallet-template build require Gradle 4.7.

msgilligan replied:
> workaround may be using `.getMajorVersion()` and do some string foo.

> a construct like `isAtLeastGradle4.7() && isJava11Compatible()`

This corresponds to "option 2" I suggested here: https://github.com/bitcoinj/bitcoinj/pull/1715#issuecomment-467086167

> This would of course make the wallet-template build require Gradle 4.7.

It looks like the OpenJFX plugin requires Gradle 4.10 or later.

schildbach commented on settings.gradle:
```diff
@@ -1,4 +1,6 @@
-include 'wallettemplate'
+if (JavaVersion.current().isJava11Compatible()) {
+    include 'wallettemplate'
+}
```
EOL missing

msgilligan replied:
Fixed and a rebased branch pushed.

schildbach commented on wallettemplate/build.gradle:
```diff
@@ -12,11 +20,16 @@ dependencies {
-    classpath = sourceSets.main.runtimeClasspath
+mainClassName = 'wallettemplate.Main'
+
+// task wallet_template and can be replaced with the 'run' task
```
This migration could actually be its own PR, since it's an independent idea that can be discussed and merged independently.

msgilligan replied:
Actually, I think it's a byproduct of the OpenJFX plugin which uses/requires the Gradle 'application' plugin.

schildbach replied:
Ah ok, didn't know. Thanks.

schildbach commented:
Perhaps we should go back to the drawing board with this. What advantages does the `org.openjfx.javafxplugin` actually have? The Gradle project page is very vague about this: 

> Plugin that makes it easy to work with JavaFX

Did you try simply adding a dependency to OpenJFX? I believe the OpenJDK dependencies are all listed here: https://mvnrepository.com/artifact/org.openjfx

msgilligan commented:
> Perhaps we should go back to the drawing board with this.

We've got (at least) three options that I can think of:

1. Try adding OpenJFX dependencies directly (as you suggest)
2. Use an alternative (non-Gradle, perhaps) mechanism to test for JDK 11+
3. Reconsider using the Gradle wrapper

> What advantages does the `org.openjfx.javafxplugin` actually have?

I'm not really sure, this is the first time I've used it and I just followed the [instructions](https://openjfx.io/openjfx-docs/#gradle). I am eager to update another JavaFX project I have to be JDK 11+, so I will be learning more about OpenJFX over the next month or two.

The documentation isn't very good about "Why", but I think the Plugin does some/all of the following:

1. Figures out which native JARs to depend on based upon the platform
2. Supports building fat jars and using the `jlink` tool to build a stripped, bundled JVM.
3. Works wit both module-based (Java 9+ module system) and classpath-based builds

I also noticed on the instructions page it says: 

> Note: Make sure to use gradle version 4.10+ and set the JAVA_HOME environment variable to JDK 11.

So this would require either using `gradlew` or checking the Gradle version before trying to include the `wallettemplate` module.

Personally, I'm a big fan of using the Gradle Wrapper. I know it breaks the rule of not checking in binaries, but it also helps significantly with ease-of-use and creating a consistent, reproducible build.

msgilligan commented:
> Don't worry about the Travis errors, I'm currently always restarting the failing jobs. Most of the time, the reason is multiple concurrently execution tests that bind to the same port.

I would really like to try to fix this. Can we turn off concurrently executing tests?

schildbach commented:
The Travis errors should be almost be fixed via 0be1008339ea0297221d378877a9c5ec45845869. I recommend rebasing again.

schildbach commented:
I'm against using Gradle wrapper. Travis uses Gradle 4.10.2 so it should be fine with the conditional. I investigated briefly but didn't find a method to programmatically check the Gradle version though.

msgilligan commented:
OK, I added a programmatic check of the Gradle version, testing with various combinations of JDK 8, 11 and Gradle 3.5.1, 4.10.2, 4.10.3, and 5.2.1 (all on macOS) and this seems to be working reasonably well.

I also added "Gradle 4.10.2 or later is recommended" to the README.

schildbach commented:
I reworded the README a bit and merged! Thanks a lot for working on this.
